contracts/fee_currencies.go

```diff
@@ -95,50 +91,52 @@ func CreditFees(
 		return err
 	}
 
-	// Solidity: function transfer(address to, uint256 amount) returns(bool)
-	transfer1Data, err := abi.Pack("transfer", feeHandler, baseTxFee)
-	if err != nil {
-		return fmt.Errorf("pack transfer base fee: %w", err)
-	}
-	_, leftoverGas, err = evm.Call(caller, *feeCurrency, transfer1Data, leftoverGas, big.NewInt(0))
+	leftoverGas, err = transferErc20(evm, abi, caller, feeCurrency, &baseFeeReceiver, baseFee, leftoverGas)
 	if err != nil {
-		return fmt.Errorf("call transfer base fee: %w", err)
+		return fmt.Errorf("base fee: %w", err)
 	}
 
-	if feeRecipient != common.ZeroAddress && tipTxFee.Cmp(common.Big0) == 1 {
-		transfer2Data, err := abi.Pack("transfer", feeRecipient, tipTxFee)
-		if err != nil {
-			return fmt.Errorf("pack transfer tip: %w", err)
-		}
-		_, leftoverGas, err = evm.Call(caller, *feeCurrency, transfer2Data, leftoverGas, big.NewInt(0))
-		if err != nil {
-			return fmt.Errorf("call transfer tip: %w", err)
-		}
-	}
-
-	if refund.Cmp(common.Big0) == 1 {
-		transfer3Data, err := abi.Pack("transfer", from, refund)
-		if err != nil {
-			return fmt.Errorf("pack transfer refund: %w", err)
-		}
-		_, leftoverGas, err = evm.Call(caller, *feeCurrency, transfer3Data, leftoverGas, big.NewInt(0))
+	if tipReceiver != common.ZeroAddress && feeTip.Cmp(common.Big0) == 1 {
+		leftoverGas, err = transferErc20(evm, abi, caller, feeCurrency, &tipReceiver, feeTip, leftoverGas)
 		if err != nil {
-			return fmt.Errorf("call transfer refund: %w", err)
+			return fmt.Errorf("fee tip: %w", err)
 		}
+	} else {
+		// If the coinbase is the zero address, instead refund the tip to the transaction sender
```

Comment:
When does this happen exactly?

Reply:
Karl and I ran into this yesterday on a simple testnet raun with `geth --dev`. So it probably doesn't happen if real validators are involved, but might happen in tests and testnets.

Reply:
In the light of the other comment, maybe worth extracting into a separate function the logic of checking for nil/0 value and zero address equality and incrementing the refund value and then calling this function early as a sort of a guardian check before the actual transfer.

Reply:
I did those changes in 8c21f53db8a636fd161df6b4640d1d7a6d5d93ee , let me know what you think.